services/device-connection/src/main/java/org/eclipse/hono/deviceconnection/infinispan/RemoteCacheBasedDeviceConnectionService.java

```diff
@@ -86,28 +84,36 @@ protected final void doStart(final Promise<Void> startPromise) {
     @Override
     protected void doStart(final Future<Void> startFuture) {
 
-        final Promise<Void> result = Promise.promise();
-        result.future().setHandler(startFuture);
-
         if (cacheManager == null) {
-            result.fail(new IllegalStateException("cache manager is not set"));
+            startFuture.fail(new IllegalStateException("cache manager is not set"));
         } else {
-            context.executeBlocking(r -> {
-                cacheManager.start();
-                cache = cacheManager.getCache("device-connection");
-                cache.start();
-                r.complete(cacheManager);
-            }, attempt -> {
-                if (attempt.succeeded()) {
-                    LOG.info("successfully connected to remote cache");
-                } else {
-                    LOG.info("failed to connect to remote cache", attempt.cause());
-                }
-            });
-            result.complete();
+            connectToGrid();
+            startFuture.complete();
         }
     }
 
+    private void connectToGrid() {
+
+        context.executeBlocking(r -> {
+            if (!cacheManager.isStarted()) {
+                log.debug("trying to start cache manager");
+                cacheManager.start();
+                log.info("started cache manager, now connecting to remote cache");
+            }
+            log.debug("trying to connect to remote cache");
+            cache = cacheManager.getCache("device-connection");
```

Comment:
If this returns `null`, then the next call will fail with a NPE. I think a more helpful exception would be appropriate here.

Reply:
I didn't know that this method can return `null`. Again, the JavaDoc doesn't seem to allow for it. However, haven taken a look at `RemoteCacheManager.getCache()` I see that it can :-1: 
Thanks for pointing out. It's a real shame that the JavaDocs of Inifispan seem to be completely useless ... I hope we are doing a much better job with Hono :-)

Reply:
We do :)